src/condor_collector.V6/collector.cpp

```
@@ -874,104 +1068,162 @@ int CollectorDaemon::receive_query_cedar_worker_thread(void *in_query_entry, Str
+		op.__mytype__ = (query_entry->adt[ix].match_mytype) ? query_entry->adt[ix].tag : nullptr;
+		op.__resultLimit__ = MIN(query_entry->limit, query_entry->adt[ix].limit) - op.__numAds__;
+		op.__results__ = &results;
+		results.clear();
+
+		CollectorHashTable * table = nullptr;
+		if (whichAds == GENERIC_AD) {
+			table = collector.getGenericHashTable(query_entry->adt[ix].tag);
+		} else if (whichAds != ANY_AD) {
+			table = collector.getHashTable(whichAds);
+		}
+		if (table) {
+			collector.walkHashTable (*table,
+				[&op](CollectorRecord*cr){
+					return op.query_scanFunc(cr);
+				});
+		} else if (ix==0 && whichAds == ANY_AD) {
+			std::vector<CollectorHashTable *> tables = collector.getAnyHashTables(op.__mytype__);
+			for (auto table : tables) {
+				collector.walkHashTable (*table,
+					[&op](CollectorRecord*cr){
+						return op.query_scanFunc(cr);
+					});
+				if (op.__numAds__ >= op.__resultLimit__)
+					break;
 			}
+			num_adtypes = 1; // don't allow Any as part of a multi-table scan.
+		} else {
+			dprintf (D_ALWAYS, "Error no collector table for %s\n", query_entry->adt[ix].tag);
+			continue;
 		}
 
-		if (evaluate_projection) {
-			proj.clear();
-			projection.clear();
-			if (EvalString(ATTR_PROJECTION, cad, curr_rec->m_publicAd, projection) && ! projection.empty()) {
+		query_time += runtime.tick(tick_time);
+
+		if (results.empty())
+			continue;
+
+		// do first time initialization for sending ads
+		if ( ! sending) {
+			// send the results via cedar
+			sock->timeout(QueryTimeout); // set up a network timeout of a longer duration
+			sock->encode();
+			sending = true;
+		}
+		if (whichAds == STARTD_PVT_AD) { filter_private_attrs = false; }
+
+		bool evaluate_projection = proj_is_expr;
+		classad::References * whitelist = proj.empty() ? nullptr : &proj;
+		if (query_entry->is_multi) {
+			std::string tagattr(query_entry->adt[ix].tag); tagattr += ATTR_PROJECTION;
+			if (query->LookupString(tagattr, projection)) {
```

If a type-specific projection attribute needs collector-side evaluation, it's silently ignored.

---

yes, type specific projections can only be strings in the current code.